extension/healthcheckv2extension/internal/http/responders.go

```diff
@@ -0,0 +1,147 @@
+// Copyright The OpenTelemetry Authors
+// SPDX-License-Identifier: Apache-2.0
+
+package http // import "github.com/open-telemetry/opentelemetry-collector-contrib/extension/healthcheckv2extension/internal/http"
+
+import (
+	"encoding/json"
+	"fmt"
+	"net/http"
+	"time"
+
+	"go.opentelemetry.io/collector/component"
+
+	"github.com/open-telemetry/opentelemetry-collector-contrib/extension/healthcheckv2extension/internal/common"
+	"github.com/open-telemetry/opentelemetry-collector-contrib/extension/healthcheckv2extension/internal/status"
+)
+
+var responseCodes = map[component.Status]int{
+	component.StatusNone:             http.StatusServiceUnavailable,
+	component.StatusStarting:         http.StatusServiceUnavailable,
+	component.StatusOK:               http.StatusOK,
+	component.StatusRecoverableError: http.StatusOK,
+	component.StatusPermanentError:   http.StatusOK,
+	component.StatusFatalError:       http.StatusInternalServerError,
+	component.StatusStopping:         http.StatusServiceUnavailable,
+	component.StatusStopped:          http.StatusServiceUnavailable,
+}
+
+type responder interface {
+	respond(*status.AggregateStatus, http.ResponseWriter)
+}
+
+type responderFunc func(*status.AggregateStatus, http.ResponseWriter)
+
+func (f responderFunc) respond(st *status.AggregateStatus, w http.ResponseWriter) {
+	f(st, w)
+}
+
+func respondWithJSON(code int, content any, w http.ResponseWriter) {
+	w.Header().Set("Content-Type", "application/json")
+	w.WriteHeader(code)
+
+	body, _ := json.Marshal(content)
+	_, _ = w.Write(body)
```

Comment:
an alternative would be to respond with an error in the json payload (and set an error status code in the response)

Reply:
my suggestion was specifically if there's an error in line 43, not so much 44

Reply:
@jpkrohling, are you suggesting I introduce a function such as:

```go
func MustWrite(w http.ResponseWriter, data []byte) {
    _, err := w.Write(data)
    if err != nil {
        panic(err)
    }
}
```

If so, I can do this, but I looked through contrib for prior art, and wasn't able to find anything similar. In fact, most call sites look like this one (e.g. `_, _ = w.Write(body)`. If you have an example of what you'd like to see that would be helpful.

Reply:
Yes, something like that. I think I'd still prefer the error to be handled, even if no (current) code path leads to that.

Reply:
Wouldn't it be better if we logged the error? A panic seems a little extreme in this case.

Reply:
+1 I think it would be better to log the error and also potentially respond with an error as Alex said above.

From the kube perspective: I think panicking here would crash the extension and crash the pod automatically in kubernetes whereas I think it would be better to handle it explicitly and return a bad response to [let the control plane handle it](https://kubernetes.io/docs/tasks/configure-pod-container/configure-liveness-readiness-startup-probes/#define-a-liveness-http-request).

Reply:
ah i see, in that case I think it would make sense to log and move on. I'm slightly panic-averse

Reply:
I made updates to handle both the situations discussed. I updated the code to log errors (at warn level) when writing to the connection. I can change this to panic if you'd prefer (cc: @jpkrohling). For the case where the response body cannot be marshaled to json, an error response body will be generated containing the reason. The response code is preserved, so the health check should succeed either way.

Reply:
I agree that a panic is an extreme, but if an error should _never_ happen (and hence, ignored), a panic wouldn't be a problem, as it would never happen :-) If we are afraid of a panic, the error should be handled.

I think the current state of the solution is great, no need to panic :-)